bauble/src/value/mod.rs

```diff
@@ -0,0 +1,723 @@
+        fn add_use_inner<C: AssetContext>(
+            this: &mut Symbols<C>,
+            leading: String,
+            end: &Spanned<PathTreeEnd>,
+        ) -> Result<()> {
+            match &end.value {
+                PathTreeEnd::Group(g) => {
+                    for node in g {
+                        let mut leading = leading.clone();
+                        for s in &node.leading.value {
+                            leading.push_str(&s.value);
+                            leading.push_str("::");
+                        }
+                        add_use_inner(this, leading, &node.end)?;
+                    }
+                }
+                PathTreeEnd::Everything => {
+                    if let Some(uses) = this.ctx.all_in(&leading) {
+                        for (ident, reference) in uses {
+                            this.add_ref(ident, reference)
+                                .map_err(|e| e.span(end.span))?;
+                        }
+                    } else {
+                        return Err(ConversionError::ModuleNotFound.span(end.span));
+                    }
+                }
+                PathTreeEnd::PathEnd(PathEnd::Ident(ident)) => {
+                    let path = format!("{leading}{ident}");
+                    if let Some(reference) = this.ctx.get_ref(&path) {
+                        this.add_ref(ident.value.clone(), reference)
+                            .map_err(|e| e.span(ident.span))?;
+                    } else {
+                        return Err(ConversionError::ModuleNotFound.span(end.span));
+                    }
+                }
+                PathTreeEnd::PathEnd(PathEnd::WithIdent(ident)) => {
+                    if let Some(reference) = this.ctx.with_ident(&leading.trim_end_matches(':'), ident.as_str()) {
+                        this.add_ref(ident.value.clone(), reference)
+                            .map_err(|e| e.span(ident.span))?;
+                    } else {
+                        return Err(ConversionError::ModuleNotFound.span(end.span));
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        let mut leading = String::new();
+        for l in use_.leading.iter() {
+            leading.push_str(l);
+            leading.push_str("::");
+        }
+        add_use_inner(self, leading, &use_.end)
+    }
+
+    fn try_resolve_macro<'a>(&'a self, ident: &str) -> Option<(&'a str, Option<Val>)> {
+        let (key, value) = self.uses.get_key_value(ident)?;
+        match value {
+            RefCopy::Unresolved => Some((key, None)),
+            RefCopy::Resolved(val) => Some((key, Some(val.clone()))),
+            RefCopy::Ref(_) => None,
+        }
+    }
+
+    fn get_module(&self, ident: &str) -> Option<&str> {
+        self.uses
+            .get(ident)
+            .and_then(|reference| reference.unwrap_ref().get_module())
+    }
+
+    fn resolve_item(&self, path: &Path) -> Result<Cow<Reference>> {
```

Comment:
I ended up rewriting this without the duplicate checks so I could understand what's going on. Here's what I came up with; you can use it if you find it more clear as well.
```rs
fn resolve_item(&self, path: &Path) -> Result<Cow<Reference>> {
        if path.leading.is_empty() {
            match &path.last.value {
                PathEnd::Ident(ident) => {
                    if let Some(asset) = self.uses.get(ident.as_str()) {
                        return Ok(Cow::Borrowed(asset.unwrap_ref()));
                    } else {
                        return Ok(Cow::Owned(
                            self.ctx
                                .get_ref(ident.as_str())
                                .ok_or(ConversionError::UnexpectedIdent.span(path.last.span))?,
                        ));
                    }
                }
                PathEnd::WithIdent(ident) => {
                    return Ok(Cow::Owned(
                        self.ctx
                            .with_ident("", ident.as_str())
                            .ok_or(ConversionError::UnexpectedIdent.span(path.last.span))?,
                    ));
                }
            }
        } else {
            let first = path.leading.first().expect("We know the Vec isn't empty.");
            let mut p = self
                .get_module(first.as_str())
                .unwrap_or(first.as_str())
                .to_string();

            for ident in &path.leading[1..] {
                p.push_str("::");
                p.push_str(ident.as_str());
            }

            return Ok(Cow::Owned(match &path.last.value {
                PathEnd::WithIdent(ident) => self
                    .ctx
                    .with_ident(&p, ident.as_str())
                    .ok_or(ConversionError::UnexpectedIdent.span(path.last.span))?,
                PathEnd::Ident(ident) => {
                    p.push_str("::");
                    p.push_str(ident.as_str());
                    self.ctx
                        .get_ref(&p)
                        .ok_or(ConversionError::UnexpectedIdent.span(path.last.span))?
                }
            }));
        }
}
```

Reply:
Did this with some minor changes